### PR TITLE
feat: Grant ownership follow up

### PR DESCRIPTION
**Pull request by sfc-gh-jcieslak**
A follow-up for #2604. 

Done in this pr:
- Add setId("") in Read (when ownership is not found on the target object) and forcefully grant ownership in Create (this was already present, but added test cases for it).
- Edge cases
  - Granting `ON PIPE` and `ON ALL PIPES` is handled (pipes are paused before and resumed after ownership transfer)

Full list of things that still need to be done:
- Deprecation messages
- More documentation (explain how grant_ownership resource handles edge cases) and examples that would show simple usage, edge cases, cases where the resource may cause trouble
- Referring to https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/2604#discussion_r1519628668, test different cases where the Delete operation may struggle with
- Test outside of Terraform interactions to see how it behaves in different situations
  - A test where used role is not privileged enough to transfer ownership
  - Also cases within Terraform to see how grant_ownership will act with other grant resources within certain configurations
- Edge cases
  - Granting `ON TASK`
  - Use `VIEW` when granting on `MATERIALIZED VIEW`
  - Granting `ON EXTERNAL TABLES`

## References
[GRANT OWNERSHIP](https://docs.snowflake.com/en/sql-reference/sql/grant-ownership)

## Mentioned in
A list of issues requesting this resource: #2549 #2199 #2084 #1942 #1875

### CODE REVIEW

File: pkg/resources/grant_ownership_acceptance_test.go [hunk 1]
```diff
@@ -577,6 +579,281 @@ func TestAcc_GrantOwnership_InvalidConfiguration_MultipleTargets(t *testing.T) {
...
+				ConfigVariables:   configVariables,
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+		},
+	})
+}
+
+func TestAcc_GrantOwnership_ForceOwnershipTransferOnCreate(t *testing.T) {
```
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_ownership_acceptance_test.go:639]: I don't understand what do we check in this test.
1. We create a role.
2. We create a Database and assign the role to it.
3. That was all preconfig. Then we have tf run which does not have to do anything as far as i understand this.

So what is checked exactly? Or what am I missing?

File: pkg/resources/grant_ownership_acceptance_test.go [hunk 2]
```diff
@@ -577,6 +579,281 @@ func TestAcc_GrantOwnership_InvalidConfiguration_MultipleTargets(t *testing.T) {
...
+					resource.TestCheckResourceAttr(resourceName, "on.0.object_name", pipeFullyQualifiedName),
+					resource.TestCheckResourceAttr(resourceName, "id", fmt.Sprintf("ToAccountRole|%s||OnObject|PIPE|%s", accountRoleFullyQualifiedName, pipeFullyQualifiedName)),
+					checkResourceOwnershipIsGranted(&sdk.ShowGrantOptions{
+						On: &sdk.ShowGrantsOn{
+							Object: &sdk.Object{
+								ObjectType: sdk.ObjectTypePipe,
+								Name:       sdk.NewSchemaObjectIdentifierFromFullyQualifiedName(pipeFullyQualifiedName),
+							},
+						},
+						// TODO(): Fix this identifier
```
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_ownership_acceptance_test.go:724]: add issue number?

File: pkg/resources/grant_ownership_acceptance_test.go [hunk 3]
```diff
@@ -577,6 +579,281 @@ func TestAcc_GrantOwnership_InvalidConfiguration_MultipleTargets(t *testing.T) {
...
+					}, sdk.ObjectTypePipe, accountRoleName, fmt.Sprintf("%s\".\"%s\".%s", acc.TestDatabaseName, acc.TestSchemaName, pipeName)),
+				),
+			},
+			{
+				ConfigDirectory:   acc.ConfigurationDirectory("TestAcc_GrantOwnership/OnPipe"),
```
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_ownership_acceptance_test.go:729]: do we need import step?

File: pkg/resources/grant_ownership_acceptance_test.go [hunk 4]
```diff
@@ -577,6 +579,281 @@ func TestAcc_GrantOwnership_InvalidConfiguration_MultipleTargets(t *testing.T) {
...
+					checkResourceOwnershipIsGranted(&sdk.ShowGrantOptions{
+						To: &sdk.ShowGrantsTo{
+							Role: sdk.NewAccountObjectIdentifier(accountRoleName),
+						},
+						// TODO: Fix this identifier
+					}, sdk.ObjectTypePipe, accountRoleName, fmt.Sprintf("%s\".\"%s\".%s", acc.TestDatabaseName, acc.TestSchemaName, pipeName), fmt.Sprintf("%s\".\"%s\".%s", acc.TestDatabaseName, acc.TestSchemaName, secondPipeName)),
+				),
+			},
+			{
+				ConfigDirectory:   acc.ConfigurationDirectory("TestAcc_GrantOwnership/OnAllPipes"),
```
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_ownership_acceptance_test.go:781]: do we need the import step?

File: pkg/resources/grant_ownership_acceptance_test.go [hunk 5]
```diff
@@ -577,6 +579,281 @@ func TestAcc_GrantOwnership_InvalidConfiguration_MultipleTargets(t *testing.T) {
...
+		new(sdk.GrantOwnershipOptions),
+	)
+	assert.NoError(t, err)
+
+	return func() {
+		assert.NoError(t, client.Databases.Drop(ctx, databaseId, &sdk.DropDatabaseOptions{}))
+	}
+}
+
+func createAccountRole(t *testing.T, name string) func() {
```
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_ownership_acceptance_test.go:820]: There is already `createAccountRoleOutsideTerraform` method in `grant_privileges_to_account_role_acceptance_test.go`, let's reuse it for now

**Review by sfc-gh-asawicki — CHANGES_REQUESTED** [review]

File: docs/resources/grant_ownership.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
+---
+page_title: "snowflake_grant_ownership Resource - terraform-provider-snowflake"
+subcategory: ""
+description: |-
+  
+---
+
+# snowflake_grant_ownership (Resource)
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply docs/resources/grant_ownership.md:8]: Yeah, I'll remove grant_ownership from the list of resources, and it'll remove itself after `make pre-push`

File: pkg/sdk/grants_impl.go [hunk 7]
```diff
@@ -226,6 +256,138 @@ func (v *grants) Show(ctx context.Context, opts *ShowGrantOptions) ([]Grant, err
 	return resultList, nil
 }
 
+// TODO (remove ?): it was only for me to grasp what has to be done in this function, but may also leave it to guide others
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/sdk/grants_impl.go:259]: Adjusted list slightly to make it render properly

File: pkg/sdk/grants_impl.go [hunk 8]
```diff
@@ -226,6 +256,138 @@ func (v *grants) Show(ctx context.Context, opts *ShowGrantOptions) ([]Grant, err
...
+
+		if err := revokeOperate(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (v *grants) grantOperateOnPipeTemporarily(ctx context.Context, pipeId SchemaObjectIdentifier, currentRole AccountObjectIdentifier) (func() error, error) {
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/sdk/grants_impl.go:353]: But still there's a case where:
1. Pipe (is running) and the current role we're using is e.g. accountadmin
2. We pause the pipe (right now, it doesn't matter how)
3. We grant ownership on pipe to role X
4. After granting ownership to role X we don't have permission on this pipe (even as accountadmin)
5. To resume the pipe we can either try to use `USE ROLE` on the role X (also hacky because it has to be granted to the user; we would have to enforce that or grant it ourselves) or grant OPERATE privilege on the current role for the time of resuming (like it is right now).

I think other options would require some additional work from the user pov. We can also just leave the pipe paused and figure something out from the perspective of pipe resource or document that pipes need manual work after ownership transfer.

File: pkg/resources/grant_ownership_acceptance_test.go [hunk 9]
```diff
@@ -577,6 +579,281 @@ func TestAcc_GrantOwnership_InvalidConfiguration_MultipleTargets(t *testing.T) {
...
+				ConfigVariables: configVariables,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "account_role_name", accountRoleName),
+					resource.TestCheckResourceAttr(resourceName, "on.0.object_type", "DATABASE"),
+					resource.TestCheckResourceAttr(resourceName, "on.0.object_name", databaseName),
+					resource.TestCheckResourceAttr(resourceName, "id", fmt.Sprintf("ToAccountRole|%s||OnObject|DATABASE|%s", accountRoleFullyQualifiedName, databaseFullyQualifiedName)),
+				),
+			},
+			{
+				ConfigDirectory:   acc.ConfigurationDirectory("TestAcc_GrantOwnership/ForceOwnershipTransferOnCreate"),
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/resources/grant_ownership_acceptance_test.go:674]: Same as above, I'll remove it.

File: pkg/resources/grant_ownership_acceptance_test.go (hunk 1 above)
Comment (sfc-gh-jcieslak) [reply pkg/resources/grant_ownership_acceptance_test.go:639]: It is supposed to show that we can transfer ownership of the objects owned by other roles than the current one (in the Create operation). Maybe this test was too much, because tbh without this working we wouldn't be able to retrieve the ownership back in the Delete operation. I can remove it.

File: pkg/resources/grant_ownership_acceptance_test.go (hunk 3 above)
Comment (sfc-gh-jcieslak) [reply pkg/resources/grant_ownership_acceptance_test.go:729]: Same as above, I'll remove it.

File: pkg/resources/grant_ownership_acceptance_test.go (hunk 4 above)
Comment (sfc-gh-jcieslak) [reply pkg/resources/grant_ownership_acceptance_test.go:781]: Same as above, I'll remove it.

File: pkg/resources/grant_ownership_acceptance_test.go (hunk 5 above)
Comment (sfc-gh-jcieslak) [reply pkg/resources/grant_ownership_acceptance_test.go:820]: Removed `createAccountRole` and used `createAccountRoleOutsideTerraform`. I was missing the register cleanup function added recently by you.

File: pkg/sdk/grants_impl.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (sfc-gh-asawicki) [reply pkg/sdk/grants_impl.go:352]: I think throwing the error there informing the user that manual work is required is the right option:
- it is not hacky
- nothing happens behind the curtains
- we knew and agreed that such cases will have to be described in the docs, so users are aware that for ownership grants manual work is expected

File: pkg/resources/grant_ownership_acceptance_test.go (hunk 1 above)
Comment (sfc-gh-asawicki) [reply pkg/resources/grant_ownership_acceptance_test.go:640]: I still don't get it. As I understand the setup (and the steps I put above), nothing happens in the create, because it is already set up correctly in the preconfig, right? So, I still don't know what this verifies.

If the delete needs to be verified, we should be explicit about it inside the test.

Comment (sfc-gh-jcieslak) [reply pkg/resources/grant_ownership_acceptance_test.go:640]: My bad, the test was wrong and I didn't see it. I changed it, so it creates one role with database owned by that role, and the second role which is used in Terraform and on Create operation "steals" ownership from the first role. It was to show what happens if we would like to transfer ownership of the resources we don't own in the first place.

File: pkg/sdk/grants_impl.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/sdk/grants_impl.go]: Changed to a non-hacky version, with as much automation as possible (it depends on the user and what privileges are granted to the pipe).

File: pkg/resources/grant_ownership_acceptance_test.go (hunk 1 above)
Comment (sfc-gh-asawicki) [reply pkg/resources/grant_ownership_acceptance_test.go:640]: okay, now it makes sense, thanks!

File: pkg/sdk/testint/grants_integration_test.go [hunk 10]
```diff
@@ -935,41 +1059,246 @@ func TestInt_GrantOwnership(t *testing.T) {
...
+
+		// Use a previously prepared role to create a pipe and grant MONITOR + OPERATE to the previously used role (ACCOUNTADMIN).
+		usePreviousRole := useRole(t, client, pipeRole.Name)
 
-		copyStatement := createPipeCopyStatement(t, table, stage)
+		pipe, pipeCleanup := createPipe(t, client, testDb(t), testSchema(t), random.AlphaN(20), copyStatement)
+		t.Cleanup(func() {
+			usePreviousRole = useRole(t, client, pipeRole.Name)
+			pipeCleanup()
+			usePreviousRole()
```
Comment (sfc-gh-asawicki) [inline pkg/sdk/testint/grants_integration_test.go:1227]: what if pipeCleanup fails? we never get to usePReviousROle, should we defer it right after `usePreviousRole = useRole(t, client, pipeRole.Name)`? The same goes for other tests too.

**Review by sfc-gh-asawicki — CHANGES_REQUESTED** [review]

File: docs/resources/grant_ownership.md [hunk 11]
```diff
@@ -0,0 +1,64 @@
+---
```
Comment (sfc-gh-asawicki) [inline docs/resources/grant_ownership.md:1]: one more thing: we should document the behavior for pipes, so that our users know how to use them. I think it would be best to add it already.

Comment (sfc-gh-jcieslak) [reply docs/resources/grant_ownership.md:1]: Added a section in the `/tmp/grant_ownership.md.tmpl`

**Review by sfc-gh-asawicki — APPROVED** [review]
